puppet/lib/puppet/util/puppetdb/char_encoding.rb

```diff
@@ -206,6 +220,137 @@ def self.get_byte(str, index)
     end
   end
 
+  # This EncodingAnalyser is supposed to be used by this module only it tries
+  # to break the string into smaller pieces in order to make line-wise diffing
+  # possible.
+  #
+  # In most circumstances the diff should give enough context to understand the
+  # issue better.
+  class EncodingAnalyser
+    class String
+      attr_reader :filename
+
+      def initialize(string, base_filename, type)
+        @string   = string
+        @type     = type
+        @filename = "#{base_filename}-#{@type}"
+        @errors   = []
+      end
+
+      def encoding
+        @string.encoding
+      end
+
+      def size
+        @string.size
+      end
+
+      def write
+        File.open(@filename, 'w') do |file|
+          file.flock(File::LOCK_EX)
+          file.puts split
+          file.chmod(0666)
+          file.flock(File::LOCK_UN)
+        end
+        filename
+      end
+
+      def errors
+        if @errors.any?
+          ["Errors in #{@type} string", @errors].join("\n")
+        else
+          ''
+        end
+      end
+
+      private
+
+      # Splitting the string does the following two things:
+      #   - split hashes and array that are values from their keys
+      #   - split lists of hashes into new lines
+      def split(second_try = false)
+        begin
+          @string.gsub(/(['"]:)({|\[)/, "\\1\n\\2").gsub(/(}|\]),({|\[)/, "\\1,\n\\2")
+        rescue ArgumentError => e
+          if second_try
+            @errors << "The #{@type} string cannot be split into diffable chunks."
+            @string
+          else
+            @string.encode!(@string.encoding, :invalid => :replace, :undef => :replace, :replace => "?")
+            @errors << "Invalid bytes encountered. These have been replaced with '?'."
+            @errors << "This should be enough to produce a helpful diff."
+            split(:second_try)
+          end
+        end
+      end
+    end
+
+    def initialize(filename)
+      @filename = filename
+    end
+
+    def original= string
+      @original = String.new(string, @filename, 'original')
+    end
+
+    def converted= string
+      @converted = String.new(string, @filename, 'converted')
+    end
+
+    def write_summary
+      File.open(@filename, 'w') do |file|
+        file.flock(File::LOCK_EX)
```

Comment:
Seems to me the flock here could be avoided completely, if we aren't overwriting the same file each time.